File: examples/live-api/paniki_client.py
#!/usr/bin/env python3
"""Talk to Pepebot's Live API with your microphone.

    python paniki_client.py                          # ws://127.0.0.1:18790/v1/live
    python paniki_client.py ws://host:18790/v1/live
    MIC="usb audio" python paniki_client.py          # pick an input device
    LANG=en-US VOICE=nenden python paniki_client.py

Speak; the agent answers and you can cut it off mid-sentence. Ctrl-C to quit.

This connects to Pepebot, not to the Realtime server directly, so the agent's
persona, skills and tools are all in play — ask it to read a file and it will.

Requires: websockets, sounddevice, numpy
"""

import asyncio
import base64
import json
import os
import queue
import sys
import threading
import urllib.request

import numpy as np
import sounddevice as sd
import websockets

sys.stdout.reconfigure(line_buffering=True)

GATEWAY = os.getenv("PEPEBOT_LIVE_URL", "ws://127.0.0.1:18790/v1/live")
PANIKI = os.getenv("PANIKI_URL", "http://100.104.36.93:8000")
LANGUAGE = os.getenv("LANG_CODE", "id-ID")
VOICE = os.getenv("VOICE", "")
# One session key per client by default, so separate clients keep separate histories.
SESSION = os.getenv("SESSION") or os.getenv("SESSION_KEY") or "py-session"

RATE = 24000        # the Realtime API speaks pcm16 mono at this rate
BLOCK = 480         # 20ms: low latency, and small enough for any frame-based VAD


class Playback:
    """Speaker buffer that can be dropped on the floor when the user cuts in."""

    def __init__(self):
        self._buffer = bytearray()
        self._lock = threading.Lock()
        self.stream = sd.RawOutputStream(
            samplerate=RATE, channels=1, dtype="int16",
            blocksize=BLOCK, callback=self._fill,
        )

    def _fill(self, outdata, frames, _time, _status):
        want = frames * 2
        with self._lock:
            chunk = bytes(self._buffer[:want])
            del self._buffer[:want]
        outdata[: len(chunk)] = chunk
        if len(chunk) < want:
            outdata[len(chunk):] = b"\x00" * (want - len(chunk))

    def add(self, pcm: bytes):
        with self._lock:
            self._buffer.extend(pcm)

    def clear(self):
        with self._lock:
            self._buffer.clear()


def discover(path: str, key: str) -> list:
    """Read /v1/models or /v1/voices so nothing here goes stale on a redeploy."""
    try:
        with urllib.request.urlopen(f"{PANIKI.rstrip('/')}{path}", timeout=8) as res:
            body = json.load(res)
        items = [row["id"] for row in body.get("data", [])]
        print(f"-- {key}: {', '.join(items[:6])}{' …' if len(items) > 6 else ''}")
        return items
    except Exception as err:
        print(f"-- could not read {path} ({err}); relying on the gateway's config")
        return []


def check_mic(device) -> None:
    """Refuse to sit there silently when the input delivers nothing.

    macOS answers a denied microphone with zeroed buffers rather than an error, so a
    permission problem is indistinguishable from a quiet room unless you look.
    """
    name = sd.query_devices(device, kind="input")["name"]
    blocks = []
    with sd.InputStream(samplerate=RATE, channels=1, dtype="int16", device=device,
                        callback=lambda indata, *_: blocks.append(indata.copy())):
        sd.sleep(600)
    peak = int(np.abs(np.concatenate(blocks)).max()) if blocks else 0
    print(f"-- mic: {name} (peak {peak})")
    if peak == 0:
        print(
            "\n!! That input is delivering pure silence, so nothing will be heard.\n"
            "   - Run this from a terminal app that holds the microphone grant.\n"
            "   - System Settings > Privacy & Security > Microphone: toggle it off and on.\n"
            "   - Wrong device? List them with:\n"
            "       python -c \"import sounddevice;print(sounddevice.query_devices())\"",
            file=sys.stderr,
        )
        sys.exit(1)


async def send_mic(ws, mic: queue.Queue):
    """Ship microphone blocks as they arrive off the audio thread."""
    sent, peak = 0, 0
    while True:
        pcm = await asyncio.to_thread(mic.get)
        peak = max(peak, int(np.abs(np.frombuffer(pcm, dtype=np.int16)).max()))
        sent += 1
        if sent % 50 == 0:  # roughly once a second
            bar = "#" * min(20, peak * 20 // 32768)
            print(f"\r  mic {peak:>5} |{bar:<20}|", end="", flush=True)
            peak = 0
        await ws.send(json.dumps({
            "type": "input_audio_buffer.append",
            "audio": base64.b64encode(pcm).decode(),
        }))


async def receive(ws, playback: Playback):
    async for message in ws:
        event = json.loads(message)

        # Pepebot's own proxy frames carry a status rather than a type.
        if event.get("status") == "connected":
            print(f"\n-- proxy: {event['provider']} -> {event['model']}")
            continue
        if event.get("status") == "reconnected":
            print("\n-- upstream reconnected; conversation state was reset")
            playback.clear()
            continue

        kind = event.get("type")

        if kind == "response.audio.delta":
            playback.add(base64.b64decode(event["delta"]))
        elif kind == "input_audio_buffer.speech_started":
            playback.clear()          # barge-in: drop what the bot had not said yet
            print("\n[listening]", end="", flush=True)
        elif kind == "conversation.item.input_audio_transcription.completed":
            print(f"\nyou: {event['transcript'].strip()}")
        elif kind == "response.audio_transcript.delta":
            print(event["delta"], end="", flush=True)
        elif kind == "response.output_item.done":
            item = event.get("item") or {}
            if item.get("type") == "function_call":
                print(f"\n[tool] {item.get('name')}({item.get('arguments', '')})"
                      " — Pepebot is running it")
        elif kind == "response.done":
            if (event.get("response") or {}).get("status") != "completed":
                playback.clear()
            print()
        elif kind == "error":
            print(f"\n[server error] {event.get('error')}")


async def main():
    url = sys.argv[1] if len(sys.argv) > 1 else GATEWAY
    device = os.getenv("MIC") or None
    if device and device.isdigit():
        device = int(device)

    models = discover("/v1/models", "models")
    discover("/v1/voices", "voices")
    check_mic(device)

    mic: queue.Queue = queue.Queue()

    def on_mic(indata, _frames, _time, _status):
        mic.put(bytes(indata))

    print(f'-- session "{SESSION}"')
    print(f"-- connecting to {url}")
    async with websockets.connect(url, max_size=None) as ws:
        setup = {
            "provider": "realtime",
            "agent": os.getenv("AGENT", "default"),
            "session_key": SESSION,
            "enable_tools": True,
            "language": LANGUAGE,
        }
        if models:
            setup["model"] = models[0]
        await ws.send(json.dumps({"setup": setup}))

        # Voice is normally set once in live.realtime_session; this is the per-session
        # override. Everything else (persona, skills, tools) Pepebot already sent.
        if VOICE:
            await ws.send(json.dumps({"type": "session.update", "session": {"voice": VOICE}}))

        playback = Playback()
        recorder = sd.RawInputStream(samplerate=RATE, channels=1, dtype="int16",
                                     blocksize=BLOCK, device=device, callback=on_mic)

        print("-- speak now (ctrl-c to quit)")
        with recorder, playback.stream:
            sender = asyncio.create_task(send_mic(ws, mic))
            try:
                await receive(ws, playback)
            finally:
                sender.cancel()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nbye")

File: examples/live-api/paniki-client.js
#!/usr/bin/env node
/**
 * Type at Pepebot's Live API, hear the answer back as a WAV file.
 *
 *   npm install ws
 *   node paniki-client.js                          # ws://127.0.0.1:18790/v1/live
 *   node paniki-client.js ws://host:18790/v1/live
 *   LANG_CODE=en-US VOICE=nenden node paniki-client.js
 *
 * Deliberately text-in / file-out: `ws` is the only dependency, so this runs
 * anywhere. For live microphone conversation use paniki_client.py, where the audio
 * bindings are far less painful.
 *
 * It connects to Pepebot rather than the Realtime server directly, so the agent's
 * persona, skills and tools are all in play — ask it to list a directory and watch
 * the tool call go by.
 */

const fs = require("fs");
const path = require("path");
const readline = require("readline");
const WebSocket = require("ws");

const GATEWAY = process.argv[2] || process.env.PEPEBOT_LIVE_URL || "ws://127.0.0.1:18790/v1/live";
const PANIKI = (process.env.PANIKI_URL || "http://100.104.36.93:8000").replace(/\/+$/, "");
const LANGUAGE = process.env.LANG_CODE || "id-ID";
const VOICE = process.env.VOICE || "";
const OUT_DIR = process.env.OUT_DIR || ".";
const RATE = 24000; // pcm16 mono, both directions

let audioChunks = [];
let turn = 0;

// Read /v1/models and /v1/voices so nothing here goes stale on a redeploy.
async function discover(pathname, label) {
  try {
    const res = await fetch(`${PANIKI}${pathname}`);
    const body = await res.json();
    const ids = (body.data || []).map((row) => row.id);
    console.log(`-- ${label}: ${ids.slice(0, 6).join(", ")}${ids.length > 6 ? " …" : ""}`);
    return ids;
  } catch (err) {
    console.log(`-- could not read ${pathname} (${err.message}); relying on the gateway's config`);
    return [];
  }
}

// pcm16 mono has no container of its own, so wrap it in a 44-byte WAV header.
function writeWav(file, pcm) {
  const header = Buffer.alloc(44);
  header.write("RIFF", 0);
  header.writeUInt32LE(36 + pcm.length, 4);
  header.write("WAVE", 8);
  header.write("fmt ", 12);
  header.writeUInt32LE(16, 16);          // PCM chunk size
  header.writeUInt16LE(1, 20);           // format: PCM
  header.writeUInt16LE(1, 22);           // channels
  header.writeUInt32LE(RATE, 24);
  header.writeUInt32LE(RATE * 2, 28);    // byte rate
  header.writeUInt16LE(2, 32);           // block align
  header.writeUInt16LE(16, 34);          // bits per sample
  header.write("data", 36);
  header.writeUInt32LE(pcm.length, 40);
  fs.writeFileSync(file, Buffer.concat([header, pcm]));
}

function flushAudio() {
  if (!audioChunks.length) return;
  const pcm = Buffer.concat(audioChunks);
  audioChunks = [];
  const file = path.join(OUT_DIR, `reply-${++turn}.wav`);
  writeWav(file, pcm);
  const seconds = (pcm.length / 2 / RATE).toFixed(1);
  console.log(`\n[audio] ${file} (${seconds}s)`);
}

async function main() {
  const models = await discover("/v1/models", "models");
  await discover("/v1/voices", "voices");

  console.log(`-- connecting to ${GATEWAY}`);
  const ws = new WebSocket(GATEWAY);

  const send = (event) => ws.send(JSON.stringify(event));

  ws.on("open", () => {
    const setup = {
      provider: "realtime",
      agent: process.env.AGENT || "default",
      session_key: process.env.SESSION_KEY || "paniki-node",
      enable_tools: true,
      language: LANGUAGE,
    };
    if (models.length) setup.model = models[0];
    send({ setup });

    // Persona, skills and tools are already on the wire from Pepebot; voice is the
    // one thing worth overriding per session.
    if (VOICE) send({ type: "session.update", session: { voice: VOICE } });
  });

  ws.on("message", (raw) => {
    let evt;
    try { evt = JSON.parse(raw.toString()); } catch { return; }

    if (evt.status === "connected") {
      console.log(`-- proxy: ${evt.provider} -> ${evt.model}`);
      console.log("-- type a message and press enter (ctrl-c to quit)");
      rl.prompt();
      return;
    }
    if (evt.status === "reconnected") {
      console.log("\n-- upstream reconnected; conversation state was reset");
      return;
    }

    switch (evt.type) {
      case "response.audio.delta":
        audioChunks.push(Buffer.from(evt.delta, "base64"));
        break;

      case "conversation.item.input_audio_transcription.completed":
        console.log(`\nyou: ${(evt.transcript || "").trim()}`);
        break;

      case "response.audio_transcript.delta":
        process.stdout.write(evt.delta || "");
        break;

      case "response.audio_transcript.done":
        console.log(`\nbot: ${(evt.transcript || "").trim()}`);
        break;

      case "response.output_item.done":
        if (evt.item && evt.item.type === "function_call") {
          console.log(`\n[tool] ${evt.item.name}(${evt.item.arguments || ""}) — Pepebot is running it`);
        }
        break;

      case "response.done":
        flushAudio();
        rl.prompt();
        break;

      case "error":
        console.log(`\n[server error] ${JSON.stringify(evt.error)}`);
        break;
    }
  });

  ws.on("error", (err) => console.error(`-- socket error: ${err.message}`));
  ws.on("close", () => { console.log("-- socket closed"); process.exit(0); });

  const rl = readline.createInterface({ input: process.stdin, output: process.stdout, prompt: "> " });
  rl.on("line", (line) => {
    const text = line.trim();
    if (!text) return rl.prompt();
    if (ws.readyState !== WebSocket.OPEN) return;
    send({
      type: "conversation.item.create",
      item: { type: "message", role: "user", content: [{ type: "input_text", text }] },
    });
    send({ type: "response.create" });
  });
  rl.on("close", () => ws.close());
}

main();
